### PR TITLE
[KEYCLOAK-7416] - Device Activity

### PR DESCRIPTION
**Pull request by pedroigor**

### CODE REVIEW

**Comment by abstractj** [comment]
@hmlnarik @stianst this PR is blocked by https://issues.jboss.org/browse/KEYCLOAK-11079

File: common/src/main/java/org/keycloak/common/Profile.java
Comment (ssilvert) [inline common/src/main/java/org/keycloak/common/Profile.java]: It was my understanding that we were concerned with the device history part so we made it experimental.  Since we have put that feature off, do we still need Device Activity to be experimental?

File: services/src/main/java/org/keycloak/services/resources/account/SessionResource.java
Comment (ssilvert) [inline services/src/main/java/org/keycloak/services/resources/account/SessionResource.java]: What does it mean for the device to have a "browser".  I would expect the browser to belong only to the session and not the device.

Also, I'm not sure we have clarified what it means for the device to have a single IP address.  Right now in the UI, I consider the device IP to be the IP from the first session in the list.  That is because this code only sets the IP for the current device, so I can't count on it being present for all devices.

**Review by ssilvert — COMMENTED** [review]
I have finished the UI using this code and it's working well.  I did have a couple of questions above, but I'm pretty happy with this code.

File: common/src/main/java/org/keycloak/common/Profile.java
Comment (pedroigor) [reply common/src/main/java/org/keycloak/common/Profile.java]: I don't have any objection to removing this feature and related checks.

File: services/src/main/java/org/keycloak/services/resources/account/SessionResource.java
Comment (pedroigor) [reply services/src/main/java/org/keycloak/services/resources/account/SessionResource.java]: The idea is that both IP and browser are available for the current device only. So you can easily check what is the browser and IP currently in use.

If you want to remove both browser and ip from device response, I have no objection.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/SessionRestServiceTest.java
Comment (pedroigor) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/SessionRestServiceTest.java]: Could you please provide more details how tests crash after setting  `oauth.setBrowserHeader("", "");`. It seems to me that when doing that you are actually breaking the webdriver and not really checking the behavior.

Did you get this?

```
org.openqa.selenium.NoSuchElementException: Unable to locate element by id for com.gargoylesoftware.htmlunit.TextPage@4e5dcc50
For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
Build info: version: '3.14.0', revision: 'aacccce0', time: '2018-08-02T20:19:58.91Z'
System info: host: 'localhost.localdomain', ip: '127.0.0.1', os.name: 'Linux', os.arch: 'amd64', os.version: '5.2.9-200.fc30.x86_64', java.version: '1.8.0_151'
Driver info: driver.version: HtmlUnitDriver
```

**Comment by pedroigor** [comment]
@ssilvert , the PR was updated to group sessions not attached with a device using an `Other` device, like we discussed.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/SessionRestServiceTest.java
Comment (abstractj) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/SessionRestServiceTest.java]: > Could you please provide more details how tests crash after setting `oauth.setBrowserHeader("", "");`. It seems to me that when doing that you are actually breaking the webdriver and not really checking the behavior.
> 
> Did you get this?

Sorry for the late reply Pedro. Yes, that's what I got into the tests. I haven't tested if it crashes the server, just brought this up to check if worth consider. If that's not the case, I think we should merge this PR.

**Comment by pedroigor** [comment]
@abstractj @ssilvert I don't know why crossdc tests are failing. They are doing fine when running locally. I've started a new job.

**Comment by abstractj** [comment]
> @abstractj @ssilvert I don't know why crossdc tests are failing. They are doing fine when running locally. I've started a new job.

That failure is completely unrelated with your changes.

File: common/src/main/java/org/keycloak/common/Profile.java
Comment (ssilvert) [reply common/src/main/java/org/keycloak/common/Profile.java]: @pedroigor Can we go ahead and remove that before we merge this PR?  

If keycloak.profile.feature.device_activity=enabled is not used then Device Activity UI will crash.  Right now it assumes that you have at least one device and one session.  If you are using the UI then that will always be true.

**Review by ssilvert — APPROVED** [review]

**Review by stianst — APPROVED** [review]
LGTM Only done a quick review though, so counting on others to do a deeper review

**Pull request merged**
